org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/NoGrammarConstraintsCodeAction.java

```diff
@@ -0,0 +1,141 @@
+			String schemaFileName = getFileName(schemaURI);
+			String schemaTemplate = generator.generate(document, sharedSettings, new XMLSchemaGeneratorSettings());
+
+			// xsi:noNamespaceSchemaLocation
+			// Create code action to create the XSD file with the generated XSD content
+			String insertText = " xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\""
+					+ document.getTextDocument().lineDelimiter(0);
+			insertText += " xsi:noNamespaceSchemaLocation=\"" + schemaFileName + "\"";
+			CodeAction noNamespaceSchemaLocationAction = createGrammarFileAndBoundIt(
+					"Bind to an XSD with xsi:noNamespaceSchemaLocation", schemaURI, schemaTemplate, insertText,
+					afterTagOffset, document, diagnostic);
+			codeActions.add(noNamespaceSchemaLocationAction);
+
+			// xml-model
+			XMLBuilder xsdWithXmlModel = new XMLBuilder(sharedSettings, null, delimiter);
+			xsdWithXmlModel.startPrologOrPI("xml-model");
+			xsdWithXmlModel.addSingleAttribute("href", schemaFileName, true);
+			xsdWithXmlModel.endPrologOrPI();
+			xsdWithXmlModel.linefeed();
+			CodeAction xsdWithXmlModelAction = createGrammarFileAndBoundIt("Bind to an XSD with xml-model", schemaURI,
+					schemaTemplate, xsdWithXmlModel.toString(), beforeTagOffset, document, diagnostic);
+			codeActions.add(xsdWithXmlModelAction);
+
+			// ---------- DTD
+
+			String dtdURI = getGrammarURI(document.getDocumentURI(), "dtd");
+			String dtdFileName = getFileName(dtdURI);
+			String dtdTemplate = generator.generate(document, sharedSettings, new DTDGeneratorSettings());
+
+			// <!DOCTYPE ${1:root-element} SYSTEM \"${2:file}.dtd\">
+			XMLBuilder docType = new XMLBuilder(sharedSettings, null, delimiter);
+			docType.startDoctype();
+			docType.addParameter(documentElement.getLocalName());
+			docType.addContent(" SYSTEM \"");
+			docType.addContent(dtdFileName);
+			docType.addContent("\"");
+			docType.endDoctype();
+			docType.linefeed();
+			CodeAction docTypeAction = createGrammarFileAndBoundIt("Bind to a DTD with DOCTYPE", dtdURI, dtdTemplate,
+					docType.toString(), beforeTagOffset, document, diagnostic);
+			codeActions.add(docTypeAction);
+
+			// xml-model
+			XMLBuilder dtdWithXmlModel = new XMLBuilder(sharedSettings, null, delimiter);
+			dtdWithXmlModel.startPrologOrPI("xml-model");
+			dtdWithXmlModel.addSingleAttribute("href", dtdFileName, true);
+			dtdWithXmlModel.endPrologOrPI();
+			dtdWithXmlModel.linefeed();
+			CodeAction dtdWithXmlModelAction = createGrammarFileAndBoundIt("Bind to a DTD with xml-model", dtdURI,
+					dtdTemplate, dtdWithXmlModel.toString(), beforeTagOffset, document, diagnostic);
+			codeActions.add(dtdWithXmlModelAction);
+
+		} catch (BadLocationException e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
+		}
+	}
+
+	private static String getGrammarURI(String documentURI, String fileExtension) {
+		int index = documentURI.lastIndexOf('.');
+		String grammarURI = documentURI.substring(0, index);

```

Comment:
Could we check for `index == -1`?

Reply:
fixed and I have improved this method to return an unique file name. I have improved too code action title to show the generated grammar file name.